c/misra/src/rules/RULE-9-4/RepeatedInitializationOfAggregateObjectElement.ql

```diff
@@ -0,0 +1,130 @@
+}
+
+/**
+ * Returns a string representation of the index of `e` relative
+ * to the nested array aggregate literal structure it is contained in.
+ */
+string getNestedArrayIndexString(Expr e) {
+  result =
+    concat(int depth |
+      depth = [0 .. getMaxDepth(getRootAggregate(e.getParent()))]
+    |
+      "[" +
+          any(int elementIndex |
+            exists(ArrayAggregateLiteral parent |
+              parent = getNthParent(e, pragma[only_bind_into](depth + 1)) and
+              parent.getElementExpr(elementIndex) = getNthParent(e, pragma[only_bind_into](depth))
+            )
+          |
+            elementIndex
+          ).toString() + "]"
+    )
+}
+
+/**
+ * Returns the number of levels of nested `ArrayAggregateLiteral`s in `al`.
+ * If there are no nested array aggregate literals, the max depth of the `ArrayAggregateLiteral` is `0`.
+ */
+language[monotonicAggregates]
+int getMaxDepth(ArrayAggregateLiteral al) {
+  if not exists(al.getElementExpr(_).(ArrayAggregateLiteral))
+  then result = 0
+  else result = 1 + max(Expr child | child = al.getElementExpr(_) | getMaxDepth(child))
+}
+
+// internal recursive predicate for `hasMultipleInitializerExprsForSameIndex`
+predicate hasMultipleInitializerExprsForSameIndexInternal(
+  ArrayAggregateLiteral al1, ArrayAggregateLiteral al2, Expr out_al1_expr, Expr out_al2_expr
+) {
+  exists(int shared_index, Expr al1_expr, Expr al2_expr |
+    // an `Expr` initializing an element of the same index in both `al1` and `al2`
+    shared_index = [0 .. al1.getArraySize() - 1] and
+    al1_expr = al1.getElementExpr(shared_index) and
+    al2_expr = al2.getElementExpr(shared_index) and
+    // but not the same `Expr`
+    not al1_expr = al2_expr and
+    (
+      // case A - the children are not aggregate literals
+      // holds if `al1` and `al2` both hold for .getElement[sharedIndex]
+      not al1_expr instanceof ArrayAggregateLiteral and
+      out_al1_expr = al1_expr and
+      out_al2_expr = al2_expr
+      or
+      // case B - `al1` and `al2` both have an aggregate literal child at the same index, so recurse
+      hasMultipleInitializerExprsForSameIndexInternal(al1_expr, al2_expr, out_al1_expr, out_al2_expr)
+    )
+  )
+}
+
+/**
+ * Holds if `expr1` and `expr2` both initialize the same array element of `root`.
+ */
+predicate hasMultipleInitializerExprsForSameIndex(ArrayAggregateLiteral root, Expr expr1, Expr expr2) {
+  hasMultipleInitializerExprsForSameIndexInternal(root, root, expr1, expr2)
+}
+
+/**
+ * Holds if `expr1` and `expr2` both initialize the same field of `root`.
+ *
+ * The dbschema keyset for `aggregate_field_init` prevents referencing multiple `Expr`
+ * that initialize the same Field and are part of the same `ClassAggregateLiteral`.
+ * This predicate is therefore unable to distinguish the individual duplicate expressions.
+ */
+predicate hasMultipleInitializerExprsForSameField(ClassAggregateLiteral root, Expr expr, Field f) {
```

Comment:
I think I dont quite understand why this then needs the `Expr` parameter, or the `expr = root.getFieldExpr(f) and` check, can it just say - there is some field initialized more than one time?

```
predicate hasMultipleInitializerExprsForSameField(ClassAggregateLiteral root, Field f) {
  count(root.getFieldExpr(f)) > 1
}
```

Reply:
Yes, that was left over from trying to identify the individual expressions. Changes made.